### ISSUE TITLE
unused_crate_dependencies false positives when dependency used only in doctests

### ISSUE THREAD

**Issue by Hawk777** [rust-lang/rust#78346]
If a dependency is only used in a doctest, it triggers an `unused_crate_dependencies` false positive.

I tried this code:

`main.rs`:
```rust
#![warn(unused_crate_dependencies)]

///! # Example
///! ```
///! use gcd::Gcd;
///! assert_eq!(6u32.gcd(8u32), 2u32);
///! ```

fn main() {
}
```

`cargo.Toml`:
```toml
[package]
name = "foo"
version = "0.1.0"
edition = "2018"

[dev-dependencies.gcd]
version = "^2.0"
```

I expected to see this happen: No warnings.

Instead, this happened: `cargo test` generates an `unused_crate_dependencies` warning for the `gcd` crate.

### Meta
`rustc --version --verbose`:
```
rustc 1.46.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.46.0
LLVM version: 10.0
```

**Comment by est31** [comment]
The `unused_crate_dependencies` lint for now isn't meant for cargo users for now, but for users of build tools like buck.

For a fix, cargo needs to be the party that emits the lint, not rustc. See my PR https://github.com/rust-lang/cargo/pull/8437

**Comment by Hawk777** [comment]
Oh, I had no idea. I saw it in the list and thought it was a reasonable way for an ordinary person writing a Rust library or program to be made aware of crates they accidentally depended on but then later stopped using. From your description in that PR, it sounds like once the warning is ready it will be enabled by default, so I should remove the `#![warn(unused_crate_dependencies)]` from my source; is that right?

**Comment by est31** [comment]
Yes, once the PR is merged, the lint itself is suppressed on the rustc side and instead rustc sends info that cargo can digest, and then cargo emits the lint instead. The level is still sent by rustc so you can still control it with adding a `#![warn(...)]` thing. Also yes, the goal is to eventually enable it by default, but I want to be a bit more careful with it and first let users opt into it. Then later once it's known that there are no bugs, it can be enabled.

You can remove the warn, or keep it in, it's your choice.

**Comment by Hawk777** [comment]
Thanks for clarifying. I think it makes sense to close this issue now.
